tests/deadcode.py

```diff
@@ -0,0 +1,269 @@
+            result = None
+            for v in node.values:
+                if self.static_value(v, deadcode) == False:
+                    result = False
+                    deadcode = True
+            return result
+
+        elif isinstance(node, ast.BoolOp) and isinstance(node.op, ast.Or):
+            result = None
+            for v in node.values:
+                if self.static_value(v, deadcode) == True:
+                    result = True
+                    deadcode = True
+            return result
+
+        elif isinstance(node, ast.Constant):
+            return bool(node.value)
+        elif isinstance(node, ast.Name) and node.id == "__debug__":
+            return True
+
+        elif isinstance(node, ast.UnaryOp) and isinstance(node.op, ast.Not):
+            v = self.static_value(node.operand, deadcode)
+            if isinstance(v, bool):
+                return not v
+        else:
+            for n in ast.iter_child_nodes(node):
+                self.walk_deadcode(n, deadcode)
+
+        return None
+
+    def check_stmts(self, stmts, deadcode):
+        """
+        used to check the body: of a function, if, ...
+        """
+        for stmt in stmts:
+            stmt.deadcode = deadcode
+
+            if self.walk_deadcode(stmt, deadcode):
+                deadcode = True
+        return deadcode
+
+    def check_childs(self, childs, deadcode):
+        """
+        used to check childs: function arguments
+        """
+        for child in childs:
+            self.walk_deadcode(child, deadcode)
+
+    def walk_annotation(self, annotation, deadcode):
+        if self.future_annotations:
+            deadcode = True
+        self.walk_deadcode(annotation, deadcode)
+
+    def walk_deadcode(self, node, deadcode):
+        "returns True if this statement will never return"
+
+        # this check is not perfect but better than nothing
+        # it tries to prevent a lot of "node without associated Bytecode" errors
+
+        # They were generated test driven.
+        # Every case represented here is derived from a error where python performed dead code elimination.
+
+        if node is None:
+            return
+
+        if isinstance(node, list):
+            for child in node:
+                self.walk_deadcode(child, deadcode)
+            return
+
+        node.deadcode = deadcode or getattr(node, "deadcode", False)
+
+        if isinstance(node, ast.Module):
+            for stmt in node.body:
+                if isinstance(stmt, ast.ImportFrom):
+                    if stmt.module == "__future__" and any(
+                        "annotations" == alias.name for alias in stmt.names
+                    ):
+                        self.future_annotations = True
+
+            self.check_stmts(node.body, deadcode)
+        elif isinstance(node, (ast.With, ast.AsyncWith)):
+            self.check_childs(node.items, deadcode)
+            deadcode = self.check_stmts(node.body, deadcode)
```

Comment:
This is incorrect, e.g:

```python
with contextlib.suppress(Exception):
    foo()
    raise Exception()
    bar()  # dead

spam()  # not dead in reality, but marked as deadcde here
```

In fact, looking at `dis` output, Python seems to treat `with` as 'catching' `return` and `continue` as if they were exceptions, so things after the `with` are still not eliminated when they should be.

Reply:
nothing propagates out of with because every operation before could raise an exception.
``` python
with contextlib.suppress(Exception):
    return 5<"5" # raises exception
no_deadcode()
```
maybe a plain return would work but this is really unlikely in the real world.

> so things after the with are still not eliminated when they should be.
I think they should not be eliminated